### PR TITLE
Improved Burn Train

### PR DESCRIPTION
**Pull request by nathanielsimard**
This PR adds many features and introduces a few breaking changes.

First, it now supports evaluation alongside training. Users can evaluate their model on multiple tasks, which are displayed in the TUI renderer.

The TUI is also improved, we now have two progress bars: one for the current task and one for the entire workflow. It also supports a new metric visualization: bar plots! This is crucial when evaluating a model on multiple datasets.

Most of the new features are tested in the MNIST training example, which was used to enhance the current inference demo model. To significantly improve training, I added the ability to compose learning rate schedulers, making it easy to implement waves of different learning rates during training, with a warmup and a decreasing learning rate to optimize training performance (gradient update scales are important).

image

image

### CODE REVIEW

File: crates/burn-core/src/lr_scheduler/composed.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+use crate::lr_scheduler::cosine::{CosineAnnealingLrScheduler, CosineAnnealingLrSchedulerConfig};
+use crate::lr_scheduler::exponential::{ExponentialLrScheduler, ExponentialLrSchedulerConfig};
+use crate::lr_scheduler::linear::{LinearLrScheduler, LinearLrSchedulerConfig};
+use crate::lr_scheduler::noam::{NoamLrScheduler, NoamLrSchedulerConfig};
+use crate::{LearningRate, config::Config};
+use burn_derive::Record;
+use burn_tensor::backend::Backend;
+
+#[derive(Config)]
+/// Compose multiple [learning rate schedulers](LrScheduler) togheter.
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/composed.rs]:
```suggestion
/// Compose multiple [learning rate schedulers](LrScheduler) together.
```

File: crates/burn-core/src/lr_scheduler/composed.rs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+pub struct ComposedLrSchedulerConfig {
+    #[config(default = "Vec::new()")]
+    schedulers: Vec<LrSchedulerConfig>,
+    #[config(default = "SchedulerReduction::Prod")]
+    reduction: SchedulerReduction,
+}
+
+#[derive(Clone)]
+/// Compose multiple [learning rate schedulers](LrScheduler) togheter.
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/composed.rs]:
```suggestion
/// Compose multiple [learning rate schedulers](LrScheduler) together.
```

File: crates/burn-core/src/lr_scheduler/composed.rs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    }
+
+    /// Appends a [linear scheduler](LinearLrScheduler).
+    pub fn linear(mut self, config: LinearLrSchedulerConfig) -> Self {
+        self.schedulers.push(LrSchedulerConfig::Linear(config));
+        self
+    }
+
+    /// Appends a [cosine scheduler](ComposedLrSchedulerConfig).
+    pub fn consine(mut self, config: CosineAnnealingLrSchedulerConfig) -> Self {
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/composed.rs]:
```suggestion
    pub fn cosine(mut self, config: CosineAnnealingLrSchedulerConfig) -> Self {
```

File: crates/burn-core/src/lr_scheduler/composed.rs [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        self.schedulers.push(LrSchedulerConfig::Cosine(config));
+        self
+    }
+
+    /// Appends a [exponential scheduler](ExponentialLrScheduler).
```
Comment (laggui) [inline crates/burn-core/src/lr_scheduler/composed.rs]:
```suggestion
    /// Appends an [exponential scheduler](ExponentialLrScheduler).
```

File: crates/burn-train/src/evaluator/builder.rs [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    renderer: Option<Box<dyn MetricsRenderer + 'static>>,
+    interrupter: Interrupter,
+    metrics: MetricsEvaluation<TO>,
+    directory: PathBuf,
+    summary: bool,
+    _p: PhantomData<(B, TI, TO)>,
+}
+
+impl<B: Backend, TI, TO: ItemLazy + 'static> EvaluatorBuilder<B, TI, TO> {
+    /// Creates a new learner builder.
```
Comment (laggui) [inline crates/burn-train/src/evaluator/builder.rs]:
```suggestion
    /// Creates a new evaluator builder.
```

File: crates/burn-train/src/evaluator/components.rs [hunk 6]
```diff
@@ -0,0 +1,53 @@
+use crate::metric::{ItemLazy, processor::EventProcessorEvaluation};
+use burn_core::{module::Module, prelude::Backend};
+use std::marker::PhantomData;
+
+/// All components necessary to train a model grouped in one trait.
```
Comment (laggui) [inline crates/burn-train/src/evaluator/components.rs]:
```suggestion
/// All components necessary to evaluate a model grouped in one trait.
```

File: crates/burn-train/src/evaluator/components.rs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+pub trait EvaluatorComponentTypes {
+    /// The backend in used for the training.
```
Comment (laggui) [inline crates/burn-train/src/evaluator/components.rs]:
```suggestion
    /// The backend in used for the evaluation.
```

File: crates/burn-train/src/evaluator/components.rs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    type Backend: Backend;
+    /// The model to train.
```
Comment (laggui) [inline crates/burn-train/src/evaluator/components.rs]:
```suggestion
    /// The model to evaluate.
```

File: crates/burn-train/src/learner/train_val.rs [hunk 9]
```diff
@@ -105,6 +106,16 @@ pub trait ValidStep<VI, VO> {
 pub(crate) type TrainLoader<LC> = Arc<dyn DataLoader<TrainBackend<LC>, InputTrain<LC>>>;
 pub(crate) type ValidLoader<LC> = Arc<dyn DataLoader<ValidBackend<LC>, InputValid<LC>>>;
 
+/// The result of a training, containing the model along with the [renderer](MetricsRenderer)..
```
Comment (laggui) [inline crates/burn-train/src/learner/train_val.rs]:
```suggestion
/// The result of a training, containing the model along with the [renderer](MetricsRenderer).
```

File: crates/burn-train/src/metric/processor/base.rs [hunk 10]
```diff
@@ -11,6 +13,14 @@ pub enum Event<T> {
     End,
 }
 
+/// Event happening during the training/validation process.
```
Comment (laggui) [inline crates/burn-train/src/metric/processor/base.rs]:
```suggestion
/// Event happening during the evaluation process.
```

File: crates/burn-train/src/metric/processor/base.rs [hunk 11]
```diff
@@ -23,16 +33,30 @@ pub trait ItemLazy: Send {
...
-    fn process_train(&mut self, event: Event<Self::ItemTrain>);
+    fn process_train(&mut self, event: LearnerEvent<Self::ItemTrain>);
     /// Collect a validation event.
-    fn process_valid(&mut self, event: Event<Self::ItemValid>);
+    fn process_valid(&mut self, event: LearnerEvent<Self::ItemValid>);
+    /// Returns the renderer used for training.
+    fn renderer(self) -> Box<dyn MetricsRenderer>;
+}
+
+/// Process events happening during training and validation.
```
Comment (laggui) [inline crates/burn-train/src/metric/processor/base.rs]:
```suggestion
/// Process events happening during evaluation.
```

File: crates/burn-train/src/metric/processor/base.rs [hunk 12]
```diff
@@ -23,16 +33,30 @@ pub trait ItemLazy: Send {
...
+pub trait EventProcessorEvaluation: Send {
+    /// The test item.
+    type ItemTest: ItemLazy;
+
+    /// Collect a training event.
```
Comment (laggui) [inline crates/burn-train/src/metric/processor/base.rs]:
```suggestion
    /// Collect a test event.
```

File: crates/burn-train/src/metric/processor/metrics.rs [hunk 13]
```diff
@@ -22,7 +38,51 @@ impl<T: ItemLazy, V: ItemLazy> Default for Metrics<T, V> {
...
+        &mut self,
+        metric: Me,
+    ) where
+        T::ItemSync: Adaptor<Me::Input> + 'static,
+    {
+        let metric = MetricWrapper::new(metric);
+        self.test_numeric.push(Box::new(metric))
+    }
+
+    /// Update the training information from the training item.
```
Comment (laggui) [inline crates/burn-train/src/metric/processor/metrics.rs]:
```suggestion
    /// Update the testing information from the testing item.
```

File: crates/burn-train/src/renderer/base.rs [hunk 14]
```diff
@@ -43,14 +44,62 @@ pub trait MetricsRenderer: Send + Sync {
...
+    pub fn new<S: core::fmt::Display>(s: S) -> Self {
+        Self {
+            name: Arc::new(format!("{s}")),
+        }
+    }
+}
+
+/// Trait for rendering metrics.
+pub trait MetricsRendererEvaluation: Send + Sync {
+    /// Updates the training metric state.
```
Comment (laggui) [inline crates/burn-train/src/renderer/base.rs]:
```suggestion
    /// Updates the testing metric state.
```

File: crates/burn-train/src/renderer/base.rs [hunk 15]
```diff
@@ -43,14 +44,62 @@ pub trait MetricsRenderer: Send + Sync {
...
+    ///
+    /// # Arguments
+    ///
+    /// * `state` - The metric state.
+    fn update_test(&mut self, name: EvaluationName, state: MetricState);
+    /// Renders the training progress.
```
Comment (laggui) [inline crates/burn-train/src/renderer/base.rs]:
```suggestion
    /// Renders the testing progress.
```

File: crates/burn-train/src/renderer/cli.rs [hunk 16]
```diff
@@ -24,3 +27,19 @@ impl MetricsRenderer for CliMetricsRenderer {
...
     }
 }
+
+impl MetricsRendererEvaluation for CliMetricsRenderer {
+    fn render_test(&mut self, item: EvaluationProgress) {
+        println!("{item:?}");
+    }
+
+    fn update_test(&mut self, _name: super::EvaluationName, _state: MetricState) {
+        //
```
Comment (laggui) [inline crates/burn-train/src/renderer/cli.rs]: Dead comment? 😄

File: crates/burn-train/src/renderer/tui/metric_numeric.rs [hunk 17]
```diff
@@ -230,6 +274,47 @@ impl NumericMetricView<'_> {
                 frame.render_widget(plot_type, chunks[1]);
                 frame.render_widget(chart, chunks[2]);
             }
+            Self::BarPlots(titles, selected, chart) => {
+                let block = Block::default()
+                    .borders(Borders::ALL)
+                    .title("Resume")
```
Comment (laggui) [inline crates/burn-train/src/renderer/tui/metric_numeric.rs]: Summary?

File: crates/burn-train/src/renderer/tui/metric_numeric.rs [hunk 18]
```diff
@@ -230,6 +274,47 @@ impl NumericMetricView<'_> {
...
+                )
+                .select(selected)
+                .style(Style::default())
+                .highlight_style(
+                    Style::default()
+                        .add_modifier(Modifier::BOLD)
+                        .add_modifier(Modifier::UNDERLINED)
+                        .fg(Color::LightYellow),
+                );
+                let title = "Resume";
```
Comment (laggui) [inline crates/burn-train/src/renderer/tui/metric_numeric.rs]: Same

File: crates/burn-train/src/renderer/base.rs [hunk 19]
```diff
@@ -43,14 +44,62 @@ pub trait MetricsRenderer: Send + Sync {
     }
 }
 
+/// A renderer that can be used for both training and evaluation.
+pub trait MetricsRenderer: MetricsRendererEvaluation + MetricsRendererTraining {
+    /// Keep the renderer from automatically closing, requiring manual action to close it.
+    fn manual_close(&mut self);
+}
```
Comment (laggui) [inline crates/burn-train/src/renderer/base.rs:51]: For the TUI we have a persistent flag that handles this iirc, can we use that instead?

Current example usage:
```rust
let learner = LearnerBuilder::new(ARTIFACT_DIR)
    // ...
    .devices(vec![device.clone()])
    .num_epochs(config.num_epochs)
    .summary();

let tui = TuiMetricsRenderer::new(learner.interrupter(), None).persistent();
let learner = learner
    .renderer(tui)
    .build(Model::new(&device), config.optimizer.init(), 1e-4);
```

(copied from https://github.com/tracel-ai/burn/issues/2233)

We could improve the pattern.

**Review by laggui — COMMENTED** [review]
Mostly naming / typos, otherwise one comment regarding persistent TUI (no automatic close / back to raw mode).

File: crates/burn-train/src/renderer/base.rs (hunk 19 above)
Comment (nathanielsimard) [reply crates/burn-train/src/renderer/base.rs:51]: I think I would remove the persistent mode at some point, with the new way, we can resize the terminal switch plot types, everything works and it's simpler, simply calling render.
